### PR TITLE
Add title for Metric.tsx

### LINKED ISSUE #1420: Add title and hover text on metric control
### Description

The `metric` control should support both the `title` and `hover_text` properties, implemented and acting just like for the `chart` control.

While working on this issue, it would be a perfect time to make 'data' typed to 'Data', and 'layout' to 'Partial&lt;Layout>'.

### Acceptance Criteria

- [ ] Ensure new code is unit tested, and check code coverage is at least 90%.
- [ ] Create related issue in taipy-doc for documentation and Release Notes.
- [ ] Check if a new demo could be provided based on this, or if legacy demos could be benefit from it.
- [ ] Ensure any change is well documented.

### Code of Conduct

- [X] I have checked the [existing issues](https://github.com/Avaiga/taipy/issues?q=is%3Aissue+).
- [ ] I am willing to work on this issue (optional)

### PR DESCRIPTION
**Pull request by namnguyen20999**
resolve https://github.com/Avaiga/taipy/issues/1420

### CODE REVIEW

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 1]
```diff
@@ -72,6 +73,9 @@ const Metric = (props: MetricProps) => {
         (delta !== undefined) && mode.push("delta");
         return [
             {
+                title: {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx]: I would have thought that title was part of the layout?

File: taipy/gui/viselements.json [hunk 2]
```diff
@@ -1116,6 +1116,11 @@
                     "shared"
                 ],
                 "properties": [
+                    {
+                        "name": "title",
+                        "type": "str",
+                        "doc": "The title of the metric."
```
Comment (FabienLelaquais) [inline taipy/gui/viselements.json]: Missing `default_value`, set to "None"
And this property must appear after "type"

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 3]
```diff
@@ -72,6 +73,9 @@ const Metric = (props: MetricProps) => {
...
+                  text: props.title
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx]: What if there's no title? See implementation in Chart.tsx - although this is based on the layout prop...

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
What about the hover_text property?

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx (hunk 1 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx]: I see that the title appear in both data and layout. I've moved it into the layout

File: taipy/gui/viselements.json (hunk 2 above)
Comment (namnguyen20999) [reply taipy/gui/viselements.json]: resolved

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx (hunk 3 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx]: I've refactored it base on the Chart.tsx

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 4]
```diff
@@ -11,7 +11,7 @@
  * specific language governing permissions and limitations under the License.
  */
 
-import React, { CSSProperties, useCallback, useEffect, useMemo, useRef, useState, lazy, Suspense } from "react";
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx:14]: It would be nice to have the same formatting rules ...
My VsCode prefers the space inside the curly brackets ;-)

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 5]
```diff
@@ -150,21 +150,21 @@ const getDecimatorsPayload = (
 ) => {
     return decimators
         ? {
-              width: plotDiv?.clientWidth,
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx:153]: idem

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 6]
```diff
@@ -616,34 +617,36 @@ const Chart = (props: ChartProp) => {
...
-                            onAfterPlot={onAfterPlot}
-                            onSelected={isOnClick(config.types) ? undefined : onSelect}
-                            onDeselect={isOnClick(config.types) ? undefined : onSelect}
-                            onClick={isOnClick(config.types) ? onSelect : undefined}
-                            config={plotConfig}
-                            useResizeHandler
-                        />
-                    )}
-                </Suspense>
+                <div>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx]: we don't need this `div`

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 7]
```diff
@@ -616,34 +617,36 @@ const Chart = (props: ChartProp) => {
     return render ? (
         <Box id={id} data-testid={props.testId} className={className} ref={plotRef}>
             <Tooltip title={hover || ""}>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx]: Just move the tooltip around the parent `Box`

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 8]
```diff
@@ -86,7 +87,7 @@ const Metric = (props: MetricProps) => {
                     prefix: extractPrefix(props.deltaFormat),
                     suffix: extractSuffix(props.deltaFormat),
                     valueformat: sprintfToD3Converter(props.deltaFormat)
-                },
+                } as Partial<Delta>,
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx:90]: is this needed if `Data` is specified ?

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 9]
```diff
@@ -144,34 +145,40 @@ const Metric = (props: MetricProps) => {
             layout.template = template;
         }
 
-        return layout
+        return {
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx]: no need to expand layout here
I propose
```
if (props.title) {
   layout.title = props.title;
}
```

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 10]
```diff
@@ -144,34 +145,40 @@ const Metric = (props: MetricProps) => {
...
         props.template_Dark_,
         props.template_Light_,
         theme.palette.mode,
-        baseLayout
+        baseLayout,
     ])
 
     return (
         <Box data-testid={props.testId} className={className}>
             <Tooltip title={hover || ""}>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx]: Move the `Tooltip` around the parent `Box`

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx [hunk 11]
```diff
@@ -144,34 +145,40 @@ const Metric = (props: MetricProps) => {
...
-                <Suspense fallback={<Skeleton key="skeleton" sx={skelStyle}/>}>
-                    <Plot
-                        data={data as Data[]}
-                        layout={layout}
-                        style={style}
-                        useResizeHandler
-                    />
-                </Suspense>
+                <div>
```
Comment (FredLL-Avaiga) [inline frontend/taipy-gui/src/components/Taipy/Metric.tsx]: no need for a div

File: taipy/gui/viselements.json [hunk 12]
```diff
@@ -1128,6 +1128,12 @@
                         "type": "str",
                         "doc": "The type of the gauge.<br/>Possible values are:\n<ul>\n<li>\"none\"</li>\n<li>\"circular\"</li>\n<li>\"linear\"</li></ul>."
                     },
+                    {
+                        "name": "title",
+                        "default_value": "None",
```
Comment (FredLL-Avaiga) [inline taipy/gui/viselements.json:1133]: @FabienLelaquais Why do we have to specify a default value when there is none

File: taipy/gui/viselements.json [hunk 13]
```diff
@@ -1192,6 +1198,11 @@
                         "name": "template[light]",
                         "type": "dict",
                         "doc": "The Plotly <a href=\"https://plotly.com/javascript/layout-template/\">layout template</a> applied over the base template when theme is not dark."
+                    },
+                    {
```
Comment (FredLL-Avaiga) [inline taipy/gui/viselements.json]: This is inherited ?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

File: taipy/gui/_renderers/factory.py [hunk 14]
```diff
@@ -360,6 +360,7 @@ class _Factory:
         .set_attributes(
             [
                 ("id",),
+                ("title",),
```
Comment (FredLL-Avaiga) [inline taipy/gui/_renderers/factory.py:363]: would it be interesting to add a render dynamic boolean property ? cf. chart

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx (hunk 10 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx]: I've moved it around the Box in the latest commit

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx (hunk 7 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Chart.tsx]: Yes I did

Comment (FabienLelaquais) [reply frontend/taipy-gui/src/components/Taipy/Chart.tsx]: ... like in Chart

File: taipy/gui/viselements.json (hunk 12 above)
Comment (FabienLelaquais) [reply taipy/gui/viselements.json:1133]: None is different from None.
Hmmm let me rephrase that.

If you don't specify a default value, then the property is required.
If you do specify a default value, even if it is "None", then the property is optional, with this value as the default.

**Review by FabienLelaquais — DISMISSED** [review]
... with Fred's comments.

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx (hunk 8 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx:90]: The prefix & suffix is not defined in the Delta interface so I have to do it this way

Comment (FredLL-Avaiga) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx:90]: OK, that's kind of strange

Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx:90]: But it does appear in Plotly js docs, I guess they forgot to implement that when converting this into React component 🧐

File: frontend/taipy-gui/src/components/Taipy/Metric.tsx (hunk 9 above)
Comment (namnguyen20999) [reply frontend/taipy-gui/src/components/Taipy/Metric.tsx]: resolved

File: taipy/gui/_renderers/factory.py (hunk 14 above)
Comment (namnguyen20999) [reply taipy/gui/_renderers/factory.py:363]: I'm not so sure what you mean here

**Review by FredLL-Avaiga — APPROVED** [review]

File: frontend/taipy-gui/src/components/Taipy/Chart.tsx [hunk 15]
```diff
@@ -417,83 +418,83 @@ const Chart = (props: ChartProp) => {
...
+                }
+                const baseZ = getValue(datum, trace, 2, true);
+                if (baseZ) {
+                    if (axisNames.length > 2) {
+                        ret[axisNames[2]] = baseZ;
+                    } else {
+                        ret.z = baseZ;
+                    }
+                }
+                    // Hack for treemap charts: create a fallback 'parents' column if needed
```
Comment (FabienLelaquais) [inline frontend/taipy-gui/src/components/Taipy/Chart.tsx:471]: So that's a strange indentation...

**Review by FabienLelaquais — APPROVED** [review]
👌
